### PR TITLE
Add a Krane CLI with one command

### PR DESCRIPTION
**Pull request by dturn**
**What are you trying to accomplish with this PR?**

Start down the road of adding a new CLI `Krane`. Our first step is adding `krane version`

See https://github.com/Shopify/kubernetes-deploy/issues/256 for lots of details

**How is this accomplished?**
We've picked `Thor` to help make the cli as easy as possible. We've picked `version` to be the first command because it's just about trivial.

**What could go wrong?**
Its new code, so the only real issue is if the dependency causes people headaches. 

**CLI**

```
➜  kubernetes-deploy git:(cli) ✗ be krane version         
[INFO][2019-08-14 11:06:56 -0700]	krane 0.26.7

➜  kubernetes-deploy git:(cli) ✗ be exe/krane  
Krane commands:
  krane help [COMMAND]  # Describe available commands or one specific command
  krane version         # Prints the version

➜  kubernetes-deploy git:(cli) ✗ be exe/krane version
[INFO][2019-08-13 16:43:39 -0700]	krane 0.26.7

➜  kubernetes-deploy git:(cli) ✗ be exe/krane version -v
ERROR: "krane version" was called with arguments ["-v"]
Usage: "krane version"
```

A fun question here is what exactly should the output be. I looked at `kubectl`, `ruby`, `node`, `bash`, and `grep` and each of these prints something different. Most are of the form `program_name version_string`. So that's what I went for.

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -20,7 +20,7 @@ desc("Run unit tests")
 Rake::TestTask.new(:unit_test) do |t|
   t.libs << "test"
   t.libs << "lib"
-  t.test_files = FileList['test/unit/**/*_test.rb']
+  t.test_files = FileList['test/unit/**/*_test.rb', 'test/exe/**/*_test.rb']
```
Comment (dturn) [inline Rakefile]: Is this a unit test or should it be a new category?

File: exe/krane [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  end
+
+  private
+
+  def logger(verbose: false)
+    @logger ||= KubernetesDeploy::FormattedLogger.build(@namespace, @context, verbose_prefix: verbose)
+  end
+end
+
+Krane.start(ARGV) if __FILE__ == $PROGRAM_NAME
```
Comment (timothysmith0609) [inline exe/krane]: Does this work? I manually bumped the gem and installed it locally and got this (added debug lines to check `__FILE__` and `$PROGRAM_NAME` when I noticed the version wouldn't print):
```
$ krane version
__FILE__:	/Users/timothysmith/.gem/ruby/2.3.3/gems/kubernetes-deploy-0.27.0/exe/krane
$PROGRAM_NAME: /Users/timothysmith/.gem/ruby/2.3.3/bin/krane
```

`bundle exec krane` yields the same result, whereas `bundle exec exe/krane version` behaves as expected.

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,12 @@
   * To see what resources may be affected, run `kubectl get $RESOURCE -o jsonpath='{ range .items[*] }{.metadata.namespace}{ "\t" }{.metadata.name}{ "\t" }{.metadata.annotations}{ "\n" }{ end }' --all-namespaces | grep "last-applied"`
   * To exclude a resource from kubernetes-deploy (and kubectl apply) management, remove the last-applied annotation `kubectl annotate $RESOURCE $SECRET_NAME kubectl.kubernetes.io/last-applied-configuration-`.
 
+*Other*
+
+- We've added a new Krane cli. This code is in alpha. We are providing
+no warranty at this time and reserve the right to make major breaking changes including
+removing it entirely at any time.
```
Comment (timothysmith0609) [inline CHANGELOG.md]: Do we have any links to provide for the Krane redesign or is all that stuff still internal? If not, it's no problem

File: Rakefile (hunk 1 above)
Comment (timothysmith0609) [reply Rakefile]: `test/cli`, perhaps?

File: exe/krane (hunk 2 above)
Comment (dturn) [reply exe/krane]: Good catch! I've re-worked the file layout,  let me know what you think.

File: lib/krane/cli/base.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    package_name "Krane"
+
+    def self.exit_on_failure?
+      true
+    end
+
+    private
+
+    def logger(verbose: false)
+      @logger ||= KubernetesDeploy::FormattedLogger.build(@namespace, @context, verbose_prefix: verbose)
```
Comment (timothysmith0609) [inline lib/krane/cli/base.rb]: Not sure if it's better to keep the base `krane` command without `@namespace, @context` and override as needed in each command or just do it all here.

**Comment by timothysmith0609** [comment]
Oops, didn't notice the failed CI before I :+1: but it's just a simple rubocop fix

File: lib/krane/cli/base.rb (hunk 4 above)
Comment (dturn) [reply lib/krane/cli/base.rb]: 🤞 this method goes away with a TaskContext object.

File: lib/krane/cli/version_command.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+require 'active_support/concern'
+
+module Krane
+  module VersionCommand
```
Comment (dturn) [inline lib/krane/cli/version_command.rb]: Given the folder structure I'm wondering if this should be `module krane; module CLI; module VersionCommand` and then have the `Krane::CLI::Krane` instead of `Krane::CLI`

Comment (timothysmith0609) [reply lib/krane/cli/version_command.rb]: I think that makes sense, since `Krane` (top-level) is the entire repo, including API accessible elements, whereas `Krane::CLI::Krane` is the actual scoping of the command-line entrypoint. It's a bit wordy but makes the separation of responsibilities clearer IMO

File: test/exe/krane_test.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    assert_logs_match_all([
+      "Krane Version: #{KubernetesDeploy::VERSION}",
+    ], in_order: true)
+  end
+
+  private
+
+  def krane
+    @krane = Krane::CLI.new.tap do |krane|
+      krane.instance_variable_set('@logger', @logger)
```
Comment (timothysmith0609) [inline test/exe/krane_test.rb]: Hmmm, why do we need to do it like this? Won't it be instantiated transparently on the call to `#logger`?

File: lib/krane/cli/version_command.rb (hunk 5 above)
Comment (dturn) [reply lib/krane/cli/version_command.rb]: Fixed

File: lib/krane/cli/version_command.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+        #check: { type: :string, banner: 'context',
+        #  desc: 'Verify if Krane is compatible with your local and remote k8s versions.' }
+      }
+
+      def self.from_options(options)
+        # This code will use the Validtor object being added in a different PR.
+        #if options[:check]
+        #  $stderr.puts("Checking if context: #{options[:check]} is valid")
+        #end
+        $stderr.puts("Krane Version: #{KubernetesDeploy::VERSION}")
```
[the remark this answers is not in the document]
Comment (dturn) [reply lib/krane/cli/version_command.rb]: Good question: https://www.jstorimer.com/blogs/workingwithcode/7766119-when-to-use-stderr-instead-of-stdout has some thoughts on each. But I think you're right, the version is the desired output of this command so it should go to stdout. (The output for the `--check` flag is probably destined for stderr)

File: lib/krane.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+require 'kubernetes-deploy/common'
```
Comment (dturn) [inline lib/krane.rb]: This is a partial optimization to make loading time faster. If we require `kubernetes-deploy` it takes about 0.8 seconds to print the `krane help`. As is about 0.4 and if we require nothing 0.3. I feel like this strikes a good balance between not feeling to slugish to print help but not have to cherry pick exactly what we want when want it.

**Comment by dturn** [comment]
This is ready for 👀 though I don't want to merge it till after the `-f` flag PR. I played with what it might look like to add the `--check` flag in https://github.com/Shopify/kubernetes-deploy/commit/e6306af47deee1776bce9d5414c36b2843b3e5f5 and feel good that there is a reasonable way forward for more than just trivial commands. @KnVerey took a stab at deploy in https://github.com/Shopify/kubernetes-deploy/compare/cli...knverey_cli and this pr took some but not all of the ideas.
